distributed/scheduler.py

```diff
@@ -2426,6 +2439,108 @@ def transition_no_worker_released(self, key, stimulus_id):
+        try:
+            ts: TaskState = self.tasks[key]
+            recommendations: dict = {}
+            client_msgs: dict = {}
+            worker_msgs: dict = {}
+
+            if self.validate:
+                assert ts not in self.queued
+                assert not self.idle, (ts, self.idle)
+                # Copied from `transition_waiting_processing`
+                assert not ts.waiting_on
+                assert not ts.who_has
+                assert not ts.exception_blame
+                assert not ts.processing_on
+                assert not ts.has_lost_dependencies
+                assert ts not in self.unrunnable
+                assert all(dts.who_has for dts in ts.dependencies)
+
+            ts.state = "queued"
+            self.queued.add(ts)
+
+            return recommendations, client_msgs, worker_msgs
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
+
+                pdb.set_trace()
+            raise
+
+    def transition_waiting_no_worker(self, key, stimulus_id):
+        try:
+            ts: TaskState = self.tasks[key]
+            recommendations: dict = {}
+            client_msgs: dict = {}
+            worker_msgs: dict = {}
+
+            if self.validate:
+                assert ts not in self.queued
+                assert not ts.who_has
+                assert not ts.exception_blame
+                assert not ts.processing_on
+                assert ts not in self.unrunnable
+
+            ts.state = "no-worker"
+            self.unrunnable.add(ts)
+
+            return recommendations, client_msgs, worker_msgs
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
+
+                pdb.set_trace()
+            raise
+
+    def transition_queued_released(self, key, stimulus_id):
+        try:
+            ts: TaskState = self.tasks[key]
+            recommendations: dict = {}
+            client_msgs: dict = {}
+            worker_msgs: dict = {}
+
+            if self.validate:
+                assert ts in self.queued
+                assert not ts.processing_on
+
+            self.queued.remove(ts)
+
+            _propagage_released(self, ts, recommendations)
+            return recommendations, client_msgs, worker_msgs
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
+
+                pdb.set_trace()
+            raise
+
+    def transition_queued_processing(self, key, stimulus_id):
+        try:
+            ts: TaskState = self.tasks[key]
+            recommendations: dict = {}
+            client_msgs: dict = {}
+            worker_msgs: dict = {}
+
+            if self.validate:
+                assert not ts.actor, f"Actors can't be queued: {ts}"
+                assert ts in self.queued
+
+            _add_to_processing(self, ts, recommendations, worker_msgs)
```

Comment:
It's correct that nothing might happen here. `queued->processing->queued` is indeed not a sane transition, but that also wouldn't happen—instead, it would just be `queued->queued`. The task state wouldn't actually ever get set to `processing`, it just might be _recommended_ to processing in instances where there are too many other tasks also being submitted for it to run.

This is leaning on the idea that recommendations are just recommendations—it can be possible that once it's time to enact them, they're no longer appropriate.

Specifically, every time a task completes and opens up a slot on a worker's threadpool, `_remove_from_processing` will peek at the next root task on the queue and recommend it to `processing`. Of course, it's also possible that when that task went to memory, one of its waiters became eligible to run. The downstream task obviously has higher priority than running a new root task. So there will be recommendations for both `downstream->processing` and `next_queued->processing`.

If it turns out we can schedule both of them, great! We will. But `downstream` gets transitioned first, since it's higher priority. If, after that, there are no longer any idle workers around, because `downstream` just took up the last open thread, then `next_queued` won't transition to `processing`—it'll stay in queued.

So maybe it's weird that sometimes this transition won't do anything, but it would be far harder to implement the alternative. If we needed to recommend `queued->processing` only when we were certain that a slot on a worker was available _and_ no other tasks with higher priority that could take that slot would also become runnable via other transitions in that cycle, `_remove_from_processing` would become very, very complicated.

With this design, `_remove_from_processing` and `_add_to_memory`—aka the selecting of queued root tasks vs the selecting of downstream tasks—can remain independent from each other.